openedx/core/djangoapps/discussions/docs/decisions/0004-in-context-discussions-linking.rst

```diff
@@ -0,0 +1,131 @@
+
+However the current mechanism has a few issues. It is stored as a JSON
+structure in the `DiscussionsIdMapping` model which has course id and a mapping
+of the discussion id to the xblock usage key in a single dict.
+
+This is OK for the existing setup because this is just a caching mechanism and
+the source of truth for this mapping is the XBlock itself, which stores the
+discussion id. On course publish this information is cached to
+`DiscussionsIdMapping`.
+
+For the new discussions system though, this mapping would be the source of
+truth for the link between discussions and units, so we should use a model
+where each such link is encoded as a row in the database.
+
+Decision
+--------
+
+Since the discussions settings need to be stored in the course structure we
+should create a new JSON structure in the course that matches the structure
+of `plugin_settings`. This can then be used to store not just the settings
+for the inbuilt discussions provider, but for any discussions provider in the
+future.
+
+When a course is published, we can copy over all the `plugin_settings` to the
+course structure with the following structure:
+
+.. code-block:: JSON
+
+    {
+        "edx-next": {
+            "discussions_enable_in_context": bool,
+            "discussions_enable_graded_units": bool,
+            "discussions_group_at_subsection": bool,
+            "discussions_custom_visibility": bool,
+        }
+    }
+
+The `edx-next` key here represents the provider id, allowing for potentially
+multiple provider configs to coexist in case of switching providers etc.
+
+To store Unit-level discussions settings, we can simply add a boolean field
+to the Unit block that specifies whether it is discussable or not. To be
+consistent with the above names we can call this field `discussions_enabled`.
+
+When any of the above settings are changed, the code handling the API call will
+go through all the units and match them against the above criterion, i.e. if
+`discussions_enable_graded_units` is false, and an existing graded unit is
+discussable (whether because an existing unit was switched to gradable, or this
+setting was), then it's `discussions_enabled` flag will be turned off.
+
+As part of the course published event, we will also traverse through all the
+Units in the course that have the flag enabled, and create the corresponding
+discussion topics in `cs_comments_service` and add a mapping. If an existing
+unit with discussions is removed, we will disable the link but not delete the
+data.
+
+The discussion grouping at subsections will simply combine the topics from all
+the units in the subsection and provide a unified view across the subsection.
+
+The mapping between discussion ids and units is also a very simple model:
+
+.. code-block:: python
+
+    class DiscussionTopicLink:
+        course_key: CourseKey
+        usage_key: UsageKey
+        discussion_provider_id: str
+        external_discussion_id: str
+        enabled: bool
```

Comment:
Do we need to store anything else in order to make it so that the course app doesn't need to check with the modulestore in the process of rendering the discussion? So for example, the title, user partition group settings, teams, etc.?

Reply:
Yes, Teams created topics via the teams topic configuration, and these are unlisted in the main discussion app topic selection they're only used to create discussions for each team discussion.